In [2]:
#!pip install mlxtend  
#! pip install --index-url https://test.pypi.org/simple/ PyARMViz

You should consider upgrading via the 'python -m pip install --upgrade pip' command.


Looking in indexes: https://test.pypi.org/simple/
  Created wheel for plotly: filename=plotly-1.7.6-cp37-none-any.whl size=118358 sha256=87c4f1d5b64a36d2b55e581dc2662e8424cd148c6ba538f6231911605377cbf0
  Stored in directory: C:\Users\classroom\AppData\Local\pip\Cache\wheels\ec\52\8c\55b910c7ce7ac04044296fb7784ee28ac4614efba73721c661
Successfully built plotly


You should consider upgrading via the 'python -m pip install --upgrade pip' command.


In [1]:
# basics
import pandas as pd
import numpy as np
import matplotlib
from matplotlib import style
import matplotlib.pyplot as plt
import seaborn as sns
from math import ceil
from collections import Counter
import datetime as dt

# market basket analysis and visualisation
from mlxtend.frequent_patterns import apriori, association_rules
from mlxtend.preprocessing import TransactionEncoder
import networkx as nx
#from PyARMViz import PyARMViz
#from PyARMViz.Rule import generate_rule_from_dict
sns.set()
#%matplotlib inline
#matplotlib.rcParams['figure.figsize'] = (18, 18)
#style.use('ggplot')


# Quarterly Analysis:
## Top Products sold

We start with analysis the top 5 products sold per store per quarter, taking into consideration units sold only. 
We will upload the 4 datasets related to the 4 different years and from there will construct a table per year that will show the top 5 products divided by quarter, together with the total of units sold for that quarter.

In [2]:
df19=pd.read_csv('2019.csv')
df=df19

In [220]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20972577 entries, 0 to 20972576
Data columns (total 10 columns):
  u         float64
  v         float64
Family      int64
Category    int64
Brand       int64
Name        int64
week        object
sku         int64
store       int64
year        int64
dtypes: float64(2), int64(7), object(1)
memory usage: 1.6+ GB


In [222]:
df['week'] = pd.to_datetime(df['week'])
df['month'] = df['week'].dt.month
df['quarter'] = df['week'].dt.quarter

In [224]:
df.head()

u       v  Family  Category  Brand  Name       week   sku  store  year  \
0  1.0  2995.0      21        34    145   291 2019-01-01  1048     69  2019   
1  2.0  2980.0      21       127    147   296 2019-01-01  1056     52  2019   
2  1.0  2850.0      21       127    147   297 2019-01-01  1057    209  2019   
3  1.0  2798.0      21       127    147   297 2019-01-01  1057     74  2019   
4  1.0  1920.0      21       127    147   297 2019-01-01  1058     51  2019   

   month  quarter  
0      1        1  
1      1        1  
2      1        1  
3      1        1  
4      1        1

In [225]:
#For 2019 we will split the dataset in quarters in order to find the top 5 products sold per store per quarter
df1=df[df['quarter']==1]
df2=df[df['quarter']==2]
df3=df[df['quarter']==3]
df4=df[df['quarter']==4]

In [226]:
#for the first quarter we select store and name of the product and we calculate the sum of units sold
quarter1 = df1.groupby(['store','Name']).agg({'  u':['sum']})
quarter1.columns = quarter1.columns.droplevel(1)
quarter1 = quarter1.sort_values(by=['store','  u'], ascending=False) #we sort Product names by units sold
quarter1 = quarter1.groupby(level=0).head(5) #we decided to show only the top 5 products for each store based on units sold

In [227]:
#for the second quarter we select store and name of the product and we calculate the sum of units sold
quarter2 = df2.groupby(['store','Name']).agg({'  u':['sum']})
quarter2.columns = quarter2.columns.droplevel(1)
quarter2 = quarter2.sort_values(by=['store','  u'], ascending=False) #we sort Product names by units sold
quarter2 = quarter2.groupby(level=0).head(5) #we decided to show only the top 5 products for each store based on units sold

In [228]:
#for the second quarter we select store and name of the product and we calculate the sum of units sold
quarter3 = df3.groupby(['store','Name']).agg({'  u':['sum']})
quarter3.columns = quarter3.columns.droplevel(1)
quarter3 = quarter3.sort_values(by=['store','  u'], ascending=False) #we sort Product names by units sold
quarter3 = quarter3.groupby(level=0).head(5) #we decided to show only the top 5 products for each store based on units sold

In [229]:
#for the second quarter we select store and name of the product and we calculate the sum of units sold
quarter4=df4.groupby(['store','Name']).agg({'  u':['sum']})
quarter4.columns = quarter4.columns.droplevel(1)
quarter4 = quarter4.sort_values(by=['store','  u'], ascending=False) #we sort Product names by units sold
quarter4 = quarter4.groupby(level=0).head(5) #we decided to show only the top 5 products for each store based on units sold

Here below an example of the results of the groupby: for stores 410 and 409 we can see the top 5 products sold per quarter taking into consideration units sold

In [230]:
quarter1.head(10)

u
store Name        
410   130   2113.0
      231   1883.0
      1855  1372.0
      1277  1333.0
      220   1166.0
409   210   1032.0
      1277   891.0
      1853   858.0
      228    835.0
      130    799.0

In [231]:
quarter2.head(10)

u
store Name        
410   231   1868.0
      1277  1475.0
      1855  1091.0
      55     980.0
      130    976.0
409   1277   951.0
      210    917.0
      1853   812.0
      228    745.0
      1786   745.0

In [232]:
quarter3.head(10)

u
store Name        
410   231   2284.0
      1277  1411.0
      1855  1243.0
      1786  1092.0
      220   1048.0
409   210    986.0
      1277   935.0
      228    863.0
      1762   758.0
      1853   735.0

In [233]:
quarter4.head(10)

u
store Name       
410   231   801.0
      130   539.0
      1277  502.0
      220   408.0
      2609  370.0
409   210   397.0
      1277  353.0
      228   273.0
      1786  250.0
      1762  246.0

In [234]:
quarter1=quarter1.reset_index()
quarter2=quarter2.reset_index()
quarter3=quarter3.reset_index()
quarter4=quarter4.reset_index()

In [235]:
#as the store repeats 4 times, in order to merge the different quarters we need first to create a unique index.
#we will use the ranking from 1st to 5th place to differentiate the stores
a = [1,2,3,4,5]
b = a*410
quarter1['ranking']=b

b1 = a*409 #quarter2-3-4 miss 1 store, so we are creating another list repeteated 409 times
quarter2['ranking']=b1
quarter3['ranking']=b1
quarter4['ranking']=b1

quarter1['Store_ranking']= quarter1['store'].astype(str)+ '.' + quarter1['ranking'].astype(str)
quarter2['Store_ranking']= quarter2['store'].astype(str)+ '.' + quarter2['ranking'].astype(str)
quarter3['Store_ranking']= quarter3['store'].astype(str)+ '.' + quarter3['ranking'].astype(str)
quarter4['Store_ranking']= quarter4['store'].astype(str)+ '.' + quarter4['ranking'].astype(str)

In [236]:
quarter1.drop(columns = ['store', 'ranking'], inplace = True)
quarter2.drop(columns = ['store', 'ranking'], inplace = True)
quarter3.drop(columns = ['store', 'ranking'], inplace = True)
quarter4.drop(columns = ['store', 'ranking'], inplace = True)

In [237]:
topproducts19=quarter1.merge(quarter2,how='outer', on='Store_ranking', suffixes=('_Q1','_Q2'))
topproducts19=topproducts19.merge(quarter3,how='outer', on='Store_ranking')
topproducts19=topproducts19.merge(quarter4, how='outer', on='Store_ranking')

In [239]:
topproducts19 = topproducts19[['Store_ranking','Name_Q1', '  u_Q1', 'Name_Q2', '  u_Q2', 'Name_x', '  u_x',
       'Name_y', '  u_y']]

In [240]:
topproducts19.columns = ['Store ranking', 'Name_Q1', 'units_Q1', 'Name_Q2', 'units_Q2', 'Name_Q3', 'units_Q3',
       'Name_Q4', 'units_Q4']

In [241]:
topproducts19

Store ranking  Name_Q1  units_Q1  Name_Q2  units_Q2  Name_Q3  units_Q3  \
0            410.1      130    2113.0    231.0    1868.0    231.0    2284.0   
1            410.2      231    1883.0   1277.0    1475.0   1277.0    1411.0   
2            410.3     1855    1372.0   1855.0    1091.0   1855.0    1243.0   
3            410.4     1277    1333.0     55.0     980.0   1786.0    1092.0   
4            410.5      220    1166.0    130.0     976.0    220.0    1048.0   
5            409.1      210    1032.0   1277.0     951.0    210.0     986.0   
6            409.2     1277     891.0    210.0     917.0   1277.0     935.0   
7            409.3     1853     858.0   1853.0     812.0    228.0     863.0   
8            409.4      228     835.0    228.0     745.0   1762.0     758.0   
9            409.5      130     799.0   1786.0     745.0   1853.0     735.0   
10           408.1      231    2352.0    231.0    2136.0    231.0    2494.0   
11           408.2      228    1038.0    228.0    1063.0    228.0    1047.0   
12           408.3     1277    1008.0   1277.0    1022.0   1277.0    1042.0   
13           408.4     1853     944.0   1853.0     933.0   1853.0     981.0   
14           408.5     1855     868.0   1059.0     767.0   1855.0     910.0   
15           407.1     1277    1545.0   1277.0    1651.0   1277.0    1652.0   
16           407.2      231    1353.0    228.0    1393.0    228.0    1448.0   
17           407.3     1853    1223.0   1853.0    1284.0    231.0    1435.0   
18           407.4     1059    1177.0   1059.0    1238.0   1853.0    1314.0   
19           407.5      228    1153.0    231.0    1228.0   1762.0    1244.0   
20           406.1      231    2194.0    231.0    2099.0    231.0    2129.0   
21           406.2     1855    1671.0   1855.0    1443.0   1855.0    1777.0   
22           406.3      559     996.0    559.0     962.0    559.0    1011.0   
23           406.4     1277     826.0   1277.0     875.0   1277.0     888.0   
24           406.5     1853     745.0   1853.0     821.0   1853.0     803.0   
25           405.1      231    3196.0    231.0    2902.0    231.0    3416.0   
26           405.2     1855    1997.0   1855.0    1890.0   1855.0    2346.0   
27           405.3     1277    1759.0   1277.0    1714.0   1277.0    1726.0   
28           405.4      559    1408.0    559.0    1440.0    559.0    1455.0   
29           405.5     1762    1298.0    228.0    1379.0   1762.0    1278.0   
...            ...      ...       ...      ...       ...      ...       ...   
2020           6.1      226     836.0    226.0     817.0    226.0     845.0   
2021           6.2     1846     786.0   1846.0     752.0   1846.0     801.0   
2022           6.3      811     651.0    811.0     628.0    811.0     638.0   
2023           6.4     1277     575.0    559.0     590.0   1277.0     591.0   
2024           6.5      559     542.0   1277.0     569.0    559.0     501.0   
2025           5.1      226    1418.0    226.0    1471.0    226.0    1530.0   
2026           5.2     1234     878.0   1846.0     699.0   1846.0     766.0   
2027           5.3     1846     664.0   1422.0     537.0   1277.0     544.0   
2028           5.4      993     543.0   1761.0     484.0   1761.0     533.0   
2029           5.5     1422     505.0   1277.0     466.0   2225.0     469.0   
2030           4.1     1234    1319.0    226.0    1047.0    226.0    1169.0   
2031           4.2      993    1069.0   1422.0    1006.0   1422.0     988.0   
2032           4.3     1422    1030.0   2609.0     890.0   2609.0     967.0   
2033           4.4      226     995.0    196.0     770.0    196.0     893.0   
2034           4.5     2609     887.0    559.0     758.0    738.0     803.0   
2035           3.1      216    1381.0    216.0    1436.0    216.0    1482.0   
2036           3.2      226     843.0   2802.0     917.0    226.0     958.0   
2037           3.3     2802     843.0    226.0     887.0   2802.0     884.0   
2038           3.4      993     766.0    198.0     804.

In [246]:
topproducts19.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2050 entries, 0 to 2049
Data columns (total 9 columns):
Store ranking    2050 non-null object
Name_Q1          2050 non-null int64
units_Q1         2050 non-null float64
Name_Q2          2045 non-null float64
units_Q2         2045 non-null float64
Name_Q3          2045 non-null float64
units_Q3         2045 non-null float64
Name_Q4          2045 non-null float64
units_Q4         2045 non-null float64
dtypes: float64(7), int64(1), object(1)
memory usage: 160.2+ KB


We will now check which is the store that was open in Q1 but then disappeared in Q2 onwards:

In [243]:
topproducts19[topproducts19['Name_Q2'].isnull()]

Store ranking  Name_Q1  units_Q1  Name_Q2  units_Q2  Name_Q3  units_Q3  \
1570          96.1      228     654.0      NaN       NaN      NaN       NaN   
1571          96.2     1234     452.0      NaN       NaN      NaN       NaN   
1572          96.3     1762     401.0      NaN       NaN      NaN       NaN   
1573          96.4      993     358.0      NaN       NaN      NaN       NaN   
1574          96.5     2609     349.0      NaN       NaN      NaN       NaN   

      Name_Q4  units_Q4  
1570      NaN       NaN  
1571      NaN       NaN  
1572      NaN       NaN  
1573      NaN       NaN  
1574      NaN       NaN

Store 96 closed after the end of the first quarter.

We will repeat the same for the other years:

## 2018

In [3]:
df18=pd.read_csv('2018.csv')
df=df18

In [182]:
df['week'] = pd.to_datetime(df['week'])
df['month'] = df['week'].dt.month
df['quarter'] = df['week'].dt.quarter
df1=df[df['quarter']==1]
df2=df[df['quarter']==2]
df3=df[df['quarter']==3]
df4=df[df['quarter']==4]

In [183]:
quarter1 = df1.groupby(['store','Name']).agg({'  u':['sum']})
quarter1.columns = quarter1.columns.droplevel(1)
quarter1 = quarter1.sort_values(by=['store','  u'], ascending=False) #we sort Product names by units sold
quarter1 = quarter1.groupby(level=0).head(5) #we decided to show only the top 5 products for each store based on units sold

#for the second quarter we select store and name of the product and we calculate the sum of units sold
quarter2 = df2.groupby(['store','Name']).agg({'  u':['sum']})
quarter2.columns = quarter2.columns.droplevel(1)
quarter2 = quarter2.sort_values(by=['store','  u'], ascending=False) #we sort Product names by units sold
quarter2 = quarter2.groupby(level=0).head(5) #we decided to show only the top 5 products for each store based on units sold

quarter3 = df3.groupby(['store','Name']).agg({'  u':['sum']})
quarter3.columns = quarter3.columns.droplevel(1)
quarter3 = quarter3.sort_values(by=['store','  u'], ascending=False) #we sort Product names by units sold
quarter3 = quarter3.groupby(level=0).head(5) 

quarter4=df4.groupby(['store','Name']).agg({'  u':['sum']})
quarter4.columns = quarter4.columns.droplevel(1)
quarter4 = quarter4.sort_values(by=['store','  u'], ascending=False) #we sort Product names by units sold
quarter4 = quarter4.groupby(level=0).head(5)

#resetting index of the groupby
quarter1=quarter1.reset_index()
quarter2=quarter2.reset_index()
quarter3=quarter3.reset_index()
quarter4=quarter4.reset_index()

In [184]:
#as the store repeats 4 times, in order to merge the different quarters we need first to create a unique index.
#we will use the ranking from 1st to 5th place to differentiate the stores
a = [1,2,3,4,5]
b = a*410
quarter1['ranking']=b
quarter2['ranking']=b
quarter3['ranking']=b
quarter4['ranking']=b


quarter1['Store_ranking']= quarter1['store'].astype(str)+ '.' + quarter1['ranking'].astype(str)
quarter2['Store_ranking']= quarter2['store'].astype(str)+ '.' + quarter2['ranking'].astype(str)
quarter3['Store_ranking']= quarter3['store'].astype(str)+ '.' + quarter3['ranking'].astype(str)
quarter4['Store_ranking']= quarter4['store'].astype(str)+ '.' + quarter4['ranking'].astype(str)

In [185]:
quarter1.drop(columns = ['store', 'ranking'], inplace = True)
quarter2.drop(columns = ['store', 'ranking'], inplace = True)
quarter3.drop(columns = ['store', 'ranking'], inplace = True)
quarter4.drop(columns = ['store', 'ranking'], inplace = True)

In [189]:
topproducts18=quarter1.merge(quarter2,how='outer', on='Store_ranking', suffixes=('_Q1','_Q2'))
topproducts18=topproducts18.merge(quarter3,how='outer', on='Store_ranking')
topproducts18=topproducts18.merge(quarter4, how='outer', on='Store_ranking')

In [190]:
topproducts18 = topproducts18[['Store_ranking','Name_Q1', '  u_Q1', 'Name_Q2', '  u_Q2', 'Name_x', '  u_x',
       'Name_y', '  u_y']]
topproducts18.columns = ['Store_ranking', 'Name_Q1', 'units_Q1', 'Name_Q2', 'units_Q2', 'Name_Q3', 'units_Q3',
       'Name_Q4', 'units_Q4']

In [191]:
topproducts18

Store_ranking  Name_Q1  units_Q1  Name_Q2  units_Q2  Name_Q3  units_Q3  \
0            410.1      130    2020.0      231    1775.0      231    1564.0   
1            410.2      231    1643.0     1277    1434.0     1277    1380.0   
2            410.3     1277    1359.0     1855    1104.0     1855    1108.0   
3            410.4     1147     966.0      220    1052.0      220     980.0   
4            410.5     1848     960.0     1786    1051.0     1786     911.0   
5            409.1     1277    1040.0     1277     944.0     1277    1000.0   
6            409.2      130     961.0      210     878.0      210     944.0   
7            409.3      210     899.0     1762     833.0     1786     779.0   
8            409.4     1762     849.0     1853     818.0     1762     746.0   
9            409.5      228     777.0      228     811.0      228     741.0   
10           408.1      231    1980.0      231    2288.5      231    2049.0   
11           408.2      228    1037.0      228     984.0      228    1053.0   
12           408.3     1277     951.0     1277     871.0     1277     933.0   
13           408.4     1853     900.0     2409     851.0     1853     875.0   
14           408.5     1855     840.0     1853     840.0     1855     875.0   
15           407.1     1277    1603.0     1277    1643.0     1277    1576.0   
16           407.2      228    1522.0      228    1446.0      231    1309.0   
17           407.3     1762    1392.0      231    1374.0      228    1250.0   
18           407.4      231    1319.0     1853    1335.0     1762    1232.0   
19           407.5     1853    1314.0     1762    1237.0     1853    1195.0   
20           406.1      231    2111.0      231    2278.0      231    2072.0   
21           406.2     1855    1654.0     1855    1624.0     1855    1665.0   
22           406.3     1277     944.0      559     944.0     1277     925.0   
23           406.4      130     902.0     1277     932.0      559     888.0   
24           406.5     1766     850.0     1766     846.0     1224     798.0   
25           405.1      231    2821.0      231    3141.0      231    2858.0   
26           405.2     1855    1920.0     1855    2049.0     1855    2130.0   
27           405.3     1277    1670.0     1277    1650.0     1277    1607.0   
28           405.4     1762    1380.0      559    1451.0      559    1421.0   
29           405.5      228    1353.0     1762    1374.0     1762    1361.0   
...            ...      ...       ...      ...       ...      ...       ...   
2020           6.1     1846     700.0     1846     748.0     1846     729.0   
2021           6.2      226     600.0      226     594.0      226     667.0   
2022           6.3      811     546.0      811     575.0      811     614.0   
2023           6.4     1277     470.0      559     515.0     1277     526.0   
2024           6.5      219     466.0      219     442.0      559     521.0   
2025           5.1      226    1143.0      226    1244.0      226    1268.0   
2026           5.2     1234     653.0     1846     592.0     1846     652.0   
2027           5.3     1846     632.0     1277     478.0     1277     437.0   
2028           5.4      993     617.0     1422     467.0      811     411.0   
2029           5.5     1422     498.0      559     386.0     1761     391.0   
2030           4.1     1234    1176.0     1422    1030.0     1422     962.0   
2031           4.2      993    1128.0      226     856.0      226     886.0   
2032           4.3     1422    1017.0     2609     849.0     2609     834.0   
2033           4.4      481     888.0      196     746.0      196     725.0   
2034           4.5      226     787.0      201     639.0      738     674.0   
2035           3.1      216    1244.0      216    1180.0      216    1184.0   
2036           3.2      993    1082.0     2802     743.0     1422     774.0   
2037           3.3      130     752.0     1422     733.0      198     764.0   
2038           3.4      198     709.0      198     730.

## 2017

In [4]:
df17=pd.read_csv('2017.csv')
df=df17

In [193]:
df['week'] = pd.to_datetime(df['week'])
df['month'] = df['week'].dt.month
df['quarter'] = df['week'].dt.quarter
df1=df[df['quarter']==1]
df2=df[df['quarter']==2]
df3=df[df['quarter']==3]
df4=df[df['quarter']==4]

In [194]:
quarter1 = df1.groupby(['store','Name']).agg({'  u':['sum']})
quarter1.columns = quarter1.columns.droplevel(1)
quarter1 = quarter1.sort_values(by=['store','  u'], ascending=False) #we sort Product names by units sold
quarter1 = quarter1.groupby(level=0).head(5) #we decided to show only the top 5 products for each store based on units sold

#for the second quarter we select store and name of the product and we calculate the sum of units sold
quarter2 = df2.groupby(['store','Name']).agg({'  u':['sum']})
quarter2.columns = quarter2.columns.droplevel(1)
quarter2 = quarter2.sort_values(by=['store','  u'], ascending=False) #we sort Product names by units sold
quarter2 = quarter2.groupby(level=0).head(5) #we decided to show only the top 5 products for each store based on units sold

quarter3 = df3.groupby(['store','Name']).agg({'  u':['sum']})
quarter3.columns = quarter3.columns.droplevel(1)
quarter3 = quarter3.sort_values(by=['store','  u'], ascending=False) #we sort Product names by units sold
quarter3 = quarter3.groupby(level=0).head(5) 

quarter4=df4.groupby(['store','Name']).agg({'  u':['sum']})
quarter4.columns = quarter4.columns.droplevel(1)
quarter4 = quarter4.sort_values(by=['store','  u'], ascending=False) #we sort Product names by units sold
quarter4 = quarter4.groupby(level=0).head(5)

#resetting index of the groupby
quarter1=quarter1.reset_index()
quarter2=quarter2.reset_index()
quarter3=quarter3.reset_index()
quarter4=quarter4.reset_index()

In [199]:
#as the store repeats 4 times, in order to merge the different quarters we need first to create a unique index.
#we will use the ranking from 1st to 5th place to differentiate the stores
a = [1,2,3,4,5]
b = a*410
quarter1['ranking']=b
quarter2['ranking']=b
quarter3['ranking']=b
quarter4['ranking']=b


quarter1['Store_ranking']= quarter1['store'].astype(str)+ '.' + quarter1['ranking'].astype(str)
quarter2['Store_ranking']= quarter2['store'].astype(str)+ '.' + quarter2['ranking'].astype(str)
quarter3['Store_ranking']= quarter3['store'].astype(str)+ '.' + quarter3['ranking'].astype(str)
quarter4['Store_ranking']= quarter4['store'].astype(str)+ '.' + quarter4['ranking'].astype(str)

In [200]:
quarter1.drop(columns = ['store', 'ranking'], inplace = True)
quarter2.drop(columns = ['store', 'ranking'], inplace = True)
quarter3.drop(columns = ['store', 'ranking'], inplace = True)
quarter4.drop(columns = ['store', 'ranking'], inplace = True)

In [201]:
topproducts17=quarter1.merge(quarter2,how='outer', on='Store_ranking', suffixes=('_Q1','_Q2'))
topproducts17=topproducts17.merge(quarter3,how='outer', on='Store_ranking')
topproducts17=topproducts17.merge(quarter4, how='outer', on='Store_ranking')

In [202]:
topproducts17 = topproducts17[['Store_ranking','Name_Q1', '  u_Q1', 'Name_Q2', '  u_Q2', 'Name_x', '  u_x',
       'Name_y', '  u_y']]
topproducts17.columns = ['Store_ranking', 'Name_Q1', 'units_Q1', 'Name_Q2', 'units_Q2', 'Name_Q3', 'units_Q3',
       'Name_Q4', 'units_Q4']

In [203]:
topproducts17

Store_ranking  Name_Q1  units_Q1  Name_Q2  units_Q2  Name_Q3  units_Q3  \
0            410.1     1287    1841.0     1277    1607.0     1277    1438.0   
1            410.2      130    1520.0      228    1486.0      228    1403.0   
2            410.3      231    1503.0      231    1346.0      231    1255.0   
3            410.4     1277    1464.0     1853    1186.0     1762    1098.0   
4            410.5      228    1275.0     1762    1122.0     1853    1083.0   
5            409.1     1277    1022.0     1277     925.0     1277     997.0   
6            409.2     1762     883.0      228     892.0      228     908.0   
7            409.3     1853     862.0      210     845.0     1853     836.0   
8            409.4      210     836.0     1786     818.0     1762     835.0   
9            409.5      228     804.0     1762     796.0     1786     766.0   
10           408.1      231    1506.5      231    1624.0      231    1667.0   
11           408.2     1277     946.0      228    1021.0      228    1067.0   
12           408.3     2409     926.0     1277     919.0     1277    1015.0   
13           408.4      228     895.0     2409     885.0     1853     847.0   
14           408.5     1762     820.0     1855     767.0     2409     788.0   
15           407.1     1277    1602.0     1277    1506.0     1762    1651.0   
16           407.2     1762    1434.0     1762    1383.0     1277    1619.0   
17           407.3      228    1388.0     1853    1262.0      228    1418.0   
18           407.4     1853    1250.0      228    1260.0     1853    1362.0   
19           407.5     2409    1082.0      231    1137.0      231    1172.0   
20           406.1      231    2196.0      231    2078.0      231    2102.0   
21           406.2     1855    1574.0     1855    1549.0     1855    1599.0   
22           406.3     1277    1066.0      307     988.0      307    1016.0   
23           406.4     2412    1059.0     2412     969.0     2412     948.0   
24           406.5      307     955.0     1277     937.0     1277     923.0   
25           405.1      231    2463.0      231    2465.0      231    2544.0   
26           405.2     1762    1846.0     1855    1822.0     1855    1857.0   
27           405.3     1855    1779.0     1762    1592.0     1762    1594.0   
28           405.4     1277    1565.0     1277    1556.0     1277    1559.0   
29           405.5      307    1392.0      228    1337.0      228    1431.0   
...            ...      ...       ...      ...       ...      ...       ...   
2020           6.1     1846     703.0     1846     664.0     1846     733.0   
2021           6.2      216     555.0     2399     498.0      216     519.0   
2022           6.3      219     528.0      811     497.0      811     517.0   
2023           6.4     2399     503.0      219     478.0      226     483.0   
2024           6.5     1277     428.0      216     465.0     2399     481.0   
2025           5.1      226     726.0      226     886.0      226    1018.0   
2026           5.2     1234     688.0     1846     613.0      567     623.0   
2027           5.3      993     578.0      567     557.0     1846     620.0   
2028           5.4      567     566.0     1422     459.0     1422     519.0   
2029           5.5     1846     559.0     1277     450.0     1277     468.0   
2030           4.1     1234    1283.0      567     894.0      567     971.0   
2031           4.2     1422     980.0     1422     849.0     1422     872.0   
2032           4.3      993     976.0      226     823.0      226     810.0   
2033           4.4      567     913.0     2609     758.0     2609     743.0   
2034           4.5      481     828.0     2685     649.0      196     692.0   
2035           3.1      216    1097.0      216    1119.0      216    1163.0   
2036           3.2      198     712.0      198     726.0      198     765.0   
2037           3.3      993     685.0      847     695.0     1422     696.0   
2038           3.4      847     682.0      912     645.

## 2016

In [5]:
df16=pd.read_csv('2016.csv')
df=df16

In [62]:
df['week'] = pd.to_datetime(df['week'])
df['month'] = df['week'].dt.month
df['quarter'] = df['week'].dt.quarter
df1=df[df['quarter']==1]
df2=df[df['quarter']==2]
df3=df[df['quarter']==3]
df4=df[df['quarter']==4]

In [63]:
quarter1 = df1.groupby(['store','Name']).agg({'  u':['sum']})
quarter1.columns = quarter1.columns.droplevel(1)
quarter1 = quarter1.sort_values(by=['store','  u'], ascending=False) #we sort Product names by units sold
quarter1 = quarter1.groupby(level=0).head(5) #we decided to show only the top 5 products for each store based on units sold

#for the second quarter we select store and name of the product and we calculate the sum of units sold
quarter2 = df2.groupby(['store','Name']).agg({'  u':['sum']})
quarter2.columns = quarter2.columns.droplevel(1)
quarter2 = quarter2.sort_values(by=['store','  u'], ascending=False) #we sort Product names by units sold
quarter2 = quarter2.groupby(level=0).head(5) #we decided to show only the top 5 products for each store based on units sold

quarter3 = df3.groupby(['store','Name']).agg({'  u':['sum']})
quarter3.columns = quarter3.columns.droplevel(1)
quarter3 = quarter3.sort_values(by=['store','  u'], ascending=False) #we sort Product names by units sold
quarter3 = quarter3.groupby(level=0).head(5) 

quarter4=df4.groupby(['store','Name']).agg({'  u':['sum']})
quarter4.columns = quarter4.columns.droplevel(1)
quarter4 = quarter4.sort_values(by=['store','  u'], ascending=False) #we sort Product names by units sold
quarter4 = quarter4.groupby(level=0).head(5)

#resetting index of the groupby
quarter1=quarter1.reset_index()
quarter2=quarter2.reset_index()
quarter3=quarter3.reset_index()
quarter4=quarter4.reset_index()

In [64]:
#as the store repeats 4 times, in order to merge the different quarters we need first to create a unique index.
#we will use the ranking from 1st to 5th place to differentiate the stores
a = [1,2,3,4,5]
b = a*410
quarter1['ranking']=b
quarter2['ranking']=b
quarter3['ranking']=b
quarter4['ranking']=b


quarter1['Store_ranking']= quarter1['store'].astype(str)+ '.' + quarter1['ranking'].astype(str)
quarter2['Store_ranking']= quarter2['store'].astype(str)+ '.' + quarter2['ranking'].astype(str)
quarter3['Store_ranking']= quarter3['store'].astype(str)+ '.' + quarter3['ranking'].astype(str)
quarter4['Store_ranking']= quarter4['store'].astype(str)+ '.' + quarter4['ranking'].astype(str)

In [212]:
quarter1.drop(columns = ['ranking'], inplace = True)
quarter2.drop(columns = ['store', 'ranking'], inplace = True)
quarter3.drop(columns = ['store', 'ranking'], inplace = True)
quarter4.drop(columns = ['store', 'ranking'], inplace = True)

In [213]:
topproducts16=quarter1.merge(quarter2,how='outer', on='Store_ranking', suffixes=('_Q1','_Q2'))
topproducts16=topproducts16.merge(quarter3,how='outer', on='Store_ranking')
topproducts16=topproducts16.merge(quarter4, how='outer', on='Store_ranking')

In [214]:
topproducts16 = topproducts16[['Store_ranking','Name_Q1', '  u_Q1', 'Name_Q2', '  u_Q2', 'Name_x', '  u_x',
       'Name_y', '  u_y']]
topproducts16.columns = ['Store', 'Name_Q1', 'units_Q1', 'Name_Q2', 'units_Q2', 'Name_Q3', 'units_Q3',
       'Name_Q4', 'units_Q4']

In [215]:
topproducts16

Store_ranking  Name_Q1  units_Q1  Name_Q2  units_Q2  Name_Q3  units_Q3  \
0            410.1     1277    1673.0      231    2048.5      231    2173.0   
1            410.2      481    1523.0     1277    1851.0     1277    1754.0   
2            410.3      231    1489.0     1762    1574.0     1762    1442.0   
3            410.4     1762    1373.0     1855    1327.0     1786    1376.0   
4            410.5      356    1315.0     1786    1276.0     1855    1212.0   
5            409.1     2409    1007.0     1277    1042.0     1277    1077.0   
6            409.2     1277     955.0     2409     994.0     1762     870.0   
7            409.3     1762     821.0      210     931.0     2409     861.0   
8            409.4     2343     812.0     1786     836.0      228     853.0   
9            409.5      210     803.0     2343     824.0     1786     811.0   
10           408.1      228    1061.0      231    1081.0      231    1196.0   
11           408.2     2409     897.0     1277     916.0      228     984.0   
12           408.3     1277     868.0      228     885.0     1277     927.0   
13           408.4     1762     754.0     2409     840.0     2409     863.0   
14           408.5      356     752.0     1762     788.0     1853     786.0   
15           407.1     1762    1482.0     1277    1512.0     1277    1602.0   
16           407.2     1277    1447.0     1762    1476.0     1762    1533.5   
17           407.3     1853    1109.0      228    1181.0      228    1290.0   
18           407.4     2409    1094.0     1853    1112.0     1853    1143.0   
19           407.5       55    1042.0     2409    1057.0       55    1069.0   
20           406.1     1855    1503.0      231    1737.5      231    1997.0   
21           406.2      231    1229.5     1855    1570.0     1855    1610.0   
22           406.3     1766    1093.0     1766    1097.0     1277    1101.0   
23           406.4     2412    1021.0     2412    1016.0     2412    1044.0   
24           406.5     1277     937.0     1277    1008.0     1224     928.0   
25           405.1     1762    2024.0     1762    2108.0      231    2001.0   
26           405.2     1855    1640.0      231    1680.0     1762    1985.0   
27           405.3     1277    1510.0     1855    1672.0     1855    1728.0   
28           405.4     1766    1387.0     1277    1532.0     1277    1652.0   
29           405.5     2412    1353.0     2412    1278.0     2412    1238.0   
...            ...      ...       ...      ...       ...      ...       ...   
2020           6.1     1846     705.0     1846     707.0     1846     728.0   
2021           6.2     2399     486.0     2399     538.0     2399     540.0   
2022           6.3      219     478.0      219     529.0      219     539.0   
2023           6.4      481     430.0      216     444.0      216     452.0   
2024           6.5     1234     411.0     2685     433.0     1277     416.0   
2025           5.1      567     652.0      567     640.0      567     602.0   
2026           5.2     1234     529.0     1846     473.0     1846     552.0   
2027           5.3     2685     464.0     2685     457.0      226     519.0   
2028           5.4      481     463.0      847     434.0     2685     446.0   
2029           5.5     1846     446.0      204     392.0      847     440.0   
2030           4.1     1234    1159.0      567    1058.0      567     965.0   
2031           4.2      567    1122.0     1422     823.0     1422     818.0   
2032           4.3      481    1030.0     2379     750.0     2379     697.0   
2033           4.4      993     903.0     2685     739.0     2685     688.0   
2034           4.5     1422     764.0     2609     696.0     2609     655.0   
2035           3.1      216     839.0      216     852.0      216     926.0   
2036           3.2      198     743.0      847     746.0      847     719.0   
2037           3.3      847     721.0      198     725.0      198     692.0   
2038           3.4      993     689.0     1422     676.

## Market share
## Families

For the market share, we will check the amount of units sold for each family for each quarter, and we will calculate the percentage of each family wrt the total units sold for that quarter.

In [3]:
df19.head()

u       v  Family  Category  Brand  Name        week   sku  store  year
0  1.0  2995.0      21        34    145   291  2019-01-01  1048     69  2019
1  2.0  2980.0      21       127    147   296  2019-01-01  1056     52  2019
2  1.0  2850.0      21       127    147   297  2019-01-01  1057    209  2019
3  1.0  2798.0      21       127    147   297  2019-01-01  1057     74  2019
4  1.0  1920.0      21       127    147   297  2019-01-01  1058     51  2019

In [7]:
df=df19

In [8]:
df['week'] = pd.to_datetime(df['week'])
df['quarter'] = df['week'].dt.quarter

In [9]:
df1=df[df['quarter']==1]
df2=df[df['quarter']==2]
df3=df[df['quarter']==3]
df4=df[df['quarter']==4]

In [11]:
# we will use a groupby to show for each store each family and the percentage of units sold
quarter1 = df1.groupby(['store','Family']).agg({'  u':['sum']})
quarter1 = quarter1.groupby(level=0).apply(lambda x:x / float(x.sum())).round(2)
quarter1.columns = quarter1.columns.droplevel(1)

In [14]:
quarter2 = df2.groupby(['store','Family']).agg({'  u':['sum']})
quarter2 = quarter2.groupby(level=0).apply(lambda x:x / float(x.sum())).round(2)
quarter2.columns = quarter2.columns.droplevel(1)

In [16]:
quarter3 = df3.groupby(['store','Family']).agg({'  u':['sum']})
quarter3 = quarter3.groupby(level=0).apply(lambda x: x / float(x.sum())).round(2)
quarter3.columns = quarter3.columns.droplevel(1)

In [17]:
quarter4 = df4.groupby(['store','Family']).agg({'  u':['sum']})
quarter4 = quarter4.groupby(level=0).apply(lambda x: x / float(x.sum())).round(2)
quarter4.columns = quarter4.columns.droplevel(1)

In [19]:
quarter1=quarter1.reset_index()
quarter2=quarter2.reset_index()
quarter3=quarter3.reset_index()
quarter4=quarter4.reset_index()

In [21]:
#as the store repeats 21 times as the number of families, in order to merge the different quarters 
#we need first to create a unique index.

quarter1['Store_family']= quarter1['store'].astype(str)+ '.' + quarter1['Family'].astype(str)
quarter2['Store_family']= quarter2['store'].astype(str)+ '.' + quarter2['Family'].astype(str)
quarter3['Store_family']= quarter3['store'].astype(str)+ '.' + quarter3['Family'].astype(str)
quarter4['Store_family']= quarter4['store'].astype(str)+ '.' + quarter4['Family'].astype(str)

In [22]:
quarter2.drop(columns = ['store', 'Family'], inplace = True)
quarter3.drop(columns = ['store', 'Family'], inplace = True)
quarter4.drop(columns = ['store', 'Family'], inplace = True)

In [23]:
marketshare19=quarter1.merge(quarter2,how='outer', on='Store_family', suffixes=('_Q1','_Q2'))
marketshare19=marketshare19.merge(quarter3,how='outer', on='Store_family')
marketshare19=marketshare19.merge(quarter4, how='outer', on='Store_family')
marketshare19.drop(columns = ['Store_family'], inplace = True)

In [26]:
#change columns name 
marketshare19.columns = ['store', 'Family', '%units_sold_Q1', '%units_sold_Q2', '%units_sold_Q3', '%units_sold_Q4']

In [27]:
marketshare19

store  Family  %units_sold_Q1  %units_sold_Q2  %units_sold_Q3  \
0       1.0     1.0            0.16            0.17            0.18   
1       1.0     2.0            0.06            0.07            0.07   
2       1.0     3.0            0.02            0.03            0.03   
3       1.0     4.0            0.00            0.00            0.00   
4       1.0     5.0            0.01            0.01            0.01   
5       1.0     6.0            0.02            0.02            0.01   
6       1.0     7.0            0.00            0.00            0.00   
7       1.0     8.0            0.03            0.04            0.04   
8       1.0     9.0            0.21            0.23            0.24   
9       1.0    10.0            0.01            0.01            0.01   
10      1.0    11.0            0.00            0.00            0.00   
11      1.0    12.0            0.15            0.18            0.17   
12      1.0    13.0            0.01            0.00            0.00   
13      1.0    14.0            0.07            0.03            0.03   
14      1.0    15.0            0.09            0.05            0.04   
15      1.0    16.0            0.02            0.01            0.01   
16      1.0    17.0            0.00            0.00            0.00   
17      1.0    18.0            0.05            0.06            0.05   
18      1.0    19.0            0.00            0.00            0.00   
19      1.0    20.0            0.00            0.00            0.00   
20      1.0    21.0            0.07            0.08            0.09   
21      2.0     1.0            0.17            0.19            0.19   
22      2.0     2.0            0.07            0.08            0.09   
23      2.0     3.0            0.03            0.03            0.03   
24      2.0     4.0            0.00            0.00            0.00   
25      2.0     5.0            0.01            0.01            0.01   
26      2.0     6.0            0.02            0.02            0.02   
27      2.0     7.0            0.01            0.01            0.01   
28      2.0     8.0            0.04            0.04            0.05   
29      2.0     9.0            0.22            0.25            0.25   
...     ...     ...             ...             ...             ...   
8579  409.0    17.0            0.00            0.00            0.00   
8580  409.0    18.0            0.07            0.08            0.08   
8581  409.0    19.0            0.01            0.01            0.01   
8582  409.0    20.0            0.00            0.00            0.00   
8583  409.0    21.0            0.07            0.08            0.07   
8584  410.0     1.0            0.17            0.17            0.17   
8585  410.0     2.0            0.06            0.07            0.07   
8586  410.0     3.0            0.02            0.02            0.03   
8587  410.0     4.0            0.00            0.00            0.00   
8588  410.0     5.0            0.01            0.01            0.01   
8589  410.0     6.0            0.01            0.01            0.01   
8590  410.0     7.0            0.01            0.00            0.01   
8591  410.0     8.0            0.02            0.02            0.02   
8592  410.0     9.0            0.17            0.19            0.19   
8593  410.0    10.0            0.01            0.01            0.01   
8594  410.0    11.0            0.00            0.00            0.00   
8595  410.0    12.0            0.20            0.21            0.22   
8596  410.0    13.0            0.01            0.00            0.00   
8597  410.0    14.0            0.10            0.05            0.03   
8598  410.0    15.0            0.05            0.03            0.02   
8599  410.0    16.0            0.02            0.01            0.01   
8600  410.0    17.0            0.00            0.00            0.00   
8601  410.0    18.0            0.06            0.06            0.07   
8602  410.0    19.0            0.01            0.01            0.01   
8603  410.0    20.0           

Let's do an example related to store 292: we will check which is the most sold Family in terms of units sold for the 1st quarter.

In [34]:
example=marketshare19[marketshare19['store']==292.0]
example.sort_values(by=['%units_sold_Q1'], ascending=False)

store  Family  %units_sold_Q1  %units_sold_Q2  %units_sold_Q3  \
6118  292.0    12.0            0.17            0.19            0.21   
6115  292.0     9.0            0.16            0.18            0.19   
6107  292.0     1.0            0.15            0.16            0.18   
6127  292.0    21.0            0.12            0.16            0.14   
6120  292.0    14.0            0.11            0.05            0.04   
6121  292.0    15.0            0.07            0.03            0.03   
6108  292.0     2.0            0.05            0.05            0.05   
6124  292.0    18.0            0.04            0.04            0.05   
6114  292.0     8.0            0.02            0.03            0.03   
6122  292.0    16.0            0.02            0.01            0.01   
6111  292.0     5.0            0.01            0.01            0.01   
6126  292.0    20.0            0.01            0.00            0.00   
6125  292.0    19.0            0.01            0.01            0.01   
6123  292.0    17.0            0.01            0.00            0.00   
6119  292.0    13.0            0.01            0.00            0.00   
6112  292.0     6.0            0.01            0.01            0.02   
6109  292.0     3.0            0.01            0.02            0.02   
6116  292.0    10.0            0.01            0.01            0.01   
6113  292.0     7.0            0.01            0.01            0.01   
6117  292.0    11.0            0.01            0.01            0.01   
6110  292.0     4.0            0.00            0.00            0.00   

      %units_sold_Q4  
6118            0.19  
6115            0.17  
6107            0.16  
6127            0.12  
6120            0.09  
6121            0.05  
6108            0.05  
6124            0.04  
6114            0.03  
6122            0.02  
6111            0.01  
6126            0.01  
6125            0.01  
6123            0.00  
6119            0.01  
6112            0.01  
6109            0.02  
6116            0.01  
6113            0.01  
6117            0.01  
6110            0.00

Family 12 accounts for 17% of the total units sold for quarter 1 for store 292, followed by family 9.
We will now repeat the same analysis for the other years.

### 2018

In [35]:
df=df18
df['week'] = pd.to_datetime(df['week'])
df['quarter'] = df['week'].dt.quarter
df1=df[df['quarter']==1]
df2=df[df['quarter']==2]
df3=df[df['quarter']==3]
df4=df[df['quarter']==4]

In [38]:
# we will use a groupby to show for each store each family and the percentage of units sold
quarter1 = df1.groupby(['store','Family']).agg({'  u':['sum']})
quarter1 = quarter1.groupby(level=0).apply(lambda x:x / float(x.sum())).round(2)
quarter1.columns = quarter1.columns.droplevel(1)

quarter2 = df2.groupby(['store','Family']).agg({'  u':['sum']})
quarter2 = quarter2.groupby(level=0).apply(lambda x:x / float(x.sum())).round(2)
quarter2.columns = quarter2.columns.droplevel(1)

quarter3 = df3.groupby(['store','Family']).agg({'  u':['sum']})
quarter3 = quarter3.groupby(level=0).apply(lambda x: x / float(x.sum())).round(2)
quarter3.columns = quarter3.columns.droplevel(1)

quarter4 = df4.groupby(['store','Family']).agg({'  u':['sum']})
quarter4 = quarter4.groupby(level=0).apply(lambda x: x / float(x.sum())).round(2)
quarter4.columns = quarter4.columns.droplevel(1)

quarter1=quarter1.reset_index()
quarter2=quarter2.reset_index()
quarter3=quarter3.reset_index()
quarter4=quarter4.reset_index()

In [43]:
#as the store repeats 21 times as the number of families, in order to merge the different quarters 
#we need first to create a unique index.

quarter1['Store_family']= quarter1['store'].astype(str)+ '.' + quarter1['Family'].astype(str)
quarter2['Store_family']= quarter2['store'].astype(str)+ '.' + quarter2['Family'].astype(str)
quarter3['Store_family']= quarter3['store'].astype(str)+ '.' + quarter3['Family'].astype(str)
quarter4['Store_family']= quarter4['store'].astype(str)+ '.' + quarter4['Family'].astype(str)

quarter2.drop(columns = ['store', 'Family'], inplace = True)
quarter3.drop(columns = ['store', 'Family'], inplace = True)
quarter4.drop(columns = ['store', 'Family'], inplace = True)

In [44]:
marketshare18=quarter1.merge(quarter2,how='outer', on='Store_family', suffixes=('_Q1','_Q2'))
marketshare18=marketshare18.merge(quarter3,how='outer', on='Store_family')
marketshare18=marketshare18.merge(quarter4, how='outer', on='Store_family')
marketshare18.drop(columns = ['Store_family'], inplace = True)

#change columns name 
marketshare18.columns = ['store', 'Family', '%units_sold_Q1', '%units_sold_Q2', '%units_sold_Q3', '%units_sold_Q4']

In [47]:
marketshare18.head(21) #let's check the behaviour with store 1

store  Family  %units_sold_Q1  %units_sold_Q2  %units_sold_Q3  \
0     1.0     1.0            0.15            0.17            0.18   
1     1.0     2.0            0.06            0.07            0.07   
2     1.0     3.0            0.02            0.02            0.02   
3     1.0     4.0            0.00            0.00            0.00   
4     1.0     5.0            0.01            0.01            0.01   
5     1.0     6.0            0.02            0.02            0.01   
6     1.0     7.0            0.01            0.01            0.01   
7     1.0     8.0            0.03            0.03            0.03   
8     1.0     9.0            0.22            0.24            0.24   
9     1.0    10.0            0.01            0.01            0.01   
10    1.0    11.0            0.00            0.00            0.00   
11    1.0    12.0            0.15            0.17            0.17   
12    1.0    13.0            0.01            0.00            0.00   
13    1.0    14.0            0.07            0.03            0.03   
14    1.0    15.0            0.10            0.05            0.03   
15    1.0    16.0            0.02            0.01            0.01   
16    1.0    17.0            0.00            0.00            0.00   
17    1.0    18.0            0.06            0.06            0.06   
18    1.0    19.0            0.00            0.00            0.00   
19    1.0    20.0            0.00            0.00            0.00   
20    1.0    21.0            0.07            0.09            0.10   

    %units_sold_Q4  
0             0.17  
1             0.06  
2             0.02  
3             0.00  
4             0.01  
5             0.02  
6             0.01  
7             0.03  
8             0.22  
9             0.01  
10            0.00  
11            0.16  
12            0.01  
13            0.07  
14            0.06  
15            0.01  
16            0.00  
17            0.06  
18            0.00  
19            0.00  
20            0.07

### 2017

In [46]:
df=df17
df['week'] = pd.to_datetime(df['week'])
df['quarter'] = df['week'].dt.quarter
df1=df[df['quarter']==1]
df2=df[df['quarter']==2]
df3=df[df['quarter']==3]
df4=df[df['quarter']==4]

In [48]:
# we will use a groupby to show for each store each family and the percentage of units sold
quarter1 = df1.groupby(['store','Family']).agg({'  u':['sum']})
quarter1 = quarter1.groupby(level=0).apply(lambda x:x / float(x.sum())).round(2)
quarter1.columns = quarter1.columns.droplevel(1)

quarter2 = df2.groupby(['store','Family']).agg({'  u':['sum']})
quarter2 = quarter2.groupby(level=0).apply(lambda x:x / float(x.sum())).round(2)
quarter2.columns = quarter2.columns.droplevel(1)

quarter3 = df3.groupby(['store','Family']).agg({'  u':['sum']})
quarter3 = quarter3.groupby(level=0).apply(lambda x: x / float(x.sum())).round(2)
quarter3.columns = quarter3.columns.droplevel(1)

quarter4 = df4.groupby(['store','Family']).agg({'  u':['sum']})
quarter4 = quarter4.groupby(level=0).apply(lambda x: x / float(x.sum())).round(2)
quarter4.columns = quarter4.columns.droplevel(1)

quarter1=quarter1.reset_index()
quarter2=quarter2.reset_index()
quarter3=quarter3.reset_index()
quarter4=quarter4.reset_index()

In [49]:
#as the store repeats 21 times as the number of families, in order to merge the different quarters 
#we need first to create a unique index.

quarter1['Store_family']= quarter1['store'].astype(str)+ '.' + quarter1['Family'].astype(str)
quarter2['Store_family']= quarter2['store'].astype(str)+ '.' + quarter2['Family'].astype(str)
quarter3['Store_family']= quarter3['store'].astype(str)+ '.' + quarter3['Family'].astype(str)
quarter4['Store_family']= quarter4['store'].astype(str)+ '.' + quarter4['Family'].astype(str)

quarter2.drop(columns = ['store', 'Family'], inplace = True)
quarter3.drop(columns = ['store', 'Family'], inplace = True)
quarter4.drop(columns = ['store', 'Family'], inplace = True)

In [50]:
marketshare17=quarter1.merge(quarter2,how='outer', on='Store_family', suffixes=('_Q1','_Q2'))
marketshare17=marketshare17.merge(quarter3,how='outer', on='Store_family')
marketshare17=marketshare17.merge(quarter4, how='outer', on='Store_family')
marketshare17.drop(columns = ['Store_family'], inplace = True)

#change columns name 
marketshare17.columns = ['store', 'Family', '%units_sold_Q1', '%units_sold_Q2', '%units_sold_Q3', '%units_sold_Q4']

In [51]:
marketshare17.head(21) #let's check the behaviour with store 1

store  Family  %units_sold_Q1  %units_sold_Q2  %units_sold_Q3  \
0     1.0     1.0            0.17            0.17            0.17   
1     1.0     2.0            0.06            0.07            0.07   
2     1.0     3.0            0.01            0.01            0.02   
3     1.0     4.0            0.00            0.00            0.00   
4     1.0     5.0            0.01            0.02            0.02   
5     1.0     6.0            0.01            0.01            0.01   
6     1.0     7.0            0.01            0.01            0.01   
7     1.0     8.0            0.02            0.02            0.02   
8     1.0     9.0            0.23            0.25            0.25   
9     1.0    10.0            0.01            0.01            0.01   
10    1.0    11.0            0.00            0.00            0.00   
11    1.0    12.0            0.15            0.17            0.17   
12    1.0    13.0            0.01            0.00            0.00   
13    1.0    14.0            0.07            0.03            0.03   
14    1.0    15.0            0.09            0.04            0.04   
15    1.0    16.0            0.02            0.01            0.01   
16    1.0    17.0            0.00             NaN             NaN   
17    1.0    18.0            0.06            0.07            0.07   
18    1.0    20.0            0.00            0.00            0.00   
19    1.0    21.0            0.07            0.10            0.11   
20    2.0     1.0            0.18            0.19            0.19   

    %units_sold_Q4  
0             0.17  
1             0.07  
2             0.02  
3             0.00  
4             0.01  
5             0.01  
6             0.01  
7             0.02  
8             0.22  
9             0.01  
10            0.00  
11            0.16  
12            0.01  
13            0.07  
14            0.07  
15            0.02  
16            0.00  
17            0.06  
18            0.00  
19            0.08  
20            0.17

### 2016

In [52]:
df=df16
df['week'] = pd.to_datetime(df['week'])
df['quarter'] = df['week'].dt.quarter
df1=df[df['quarter']==1]
df2=df[df['quarter']==2]
df3=df[df['quarter']==3]
df4=df[df['quarter']==4]

In [53]:
# we will use a groupby to show for each store each family and the percentage of units sold
quarter1 = df1.groupby(['store','Family']).agg({'  u':['sum']})
quarter1 = quarter1.groupby(level=0).apply(lambda x:x / float(x.sum())).round(2)
quarter1.columns = quarter1.columns.droplevel(1)

quarter2 = df2.groupby(['store','Family']).agg({'  u':['sum']})
quarter2 = quarter2.groupby(level=0).apply(lambda x:x / float(x.sum())).round(2)
quarter2.columns = quarter2.columns.droplevel(1)

quarter3 = df3.groupby(['store','Family']).agg({'  u':['sum']})
quarter3 = quarter3.groupby(level=0).apply(lambda x: x / float(x.sum())).round(2)
quarter3.columns = quarter3.columns.droplevel(1)

quarter4 = df4.groupby(['store','Family']).agg({'  u':['sum']})
quarter4 = quarter4.groupby(level=0).apply(lambda x: x / float(x.sum())).round(2)
quarter4.columns = quarter4.columns.droplevel(1)

quarter1=quarter1.reset_index()
quarter2=quarter2.reset_index()
quarter3=quarter3.reset_index()
quarter4=quarter4.reset_index()

In [54]:
#as the store repeats 21 times as the number of families, in order to merge the different quarters 
#we need first to create a unique index.

quarter1['Store_family']= quarter1['store'].astype(str)+ '.' + quarter1['Family'].astype(str)
quarter2['Store_family']= quarter2['store'].astype(str)+ '.' + quarter2['Family'].astype(str)
quarter3['Store_family']= quarter3['store'].astype(str)+ '.' + quarter3['Family'].astype(str)
quarter4['Store_family']= quarter4['store'].astype(str)+ '.' + quarter4['Family'].astype(str)

quarter2.drop(columns = ['store', 'Family'], inplace = True)
quarter3.drop(columns = ['store', 'Family'], inplace = True)
quarter4.drop(columns = ['store', 'Family'], inplace = True)

In [55]:
marketshare16=quarter1.merge(quarter2,how='outer', on='Store_family', suffixes=('_Q1','_Q2'))
marketshare16=marketshare16.merge(quarter3,how='outer', on='Store_family')
marketshare16=marketshare16.merge(quarter4, how='outer', on='Store_family')
marketshare16.drop(columns = ['Store_family'], inplace = True)

#change columns name 
marketshare16.columns = ['store', 'Family', '%units_sold_Q1', '%units_sold_Q2', '%units_sold_Q3', '%units_sold_Q4']

In [57]:
marketshare16.head(21)

store  Family  %units_sold_Q1  %units_sold_Q2  %units_sold_Q3  \
0     1.0     1.0            0.17            0.17            0.18   
1     1.0     2.0            0.06            0.07            0.07   
2     1.0     3.0            0.01            0.01            0.01   
3     1.0     4.0            0.00            0.00            0.00   
4     1.0     5.0            0.02            0.02            0.02   
5     1.0     6.0            0.02            0.02            0.01   
6     1.0     7.0            0.01            0.01            0.01   
7     1.0     8.0            0.01            0.01            0.01   
8     1.0     9.0            0.23            0.24            0.25   
9     1.0    10.0            0.01            0.01            0.01   
10    1.0    11.0            0.00            0.00            0.00   
11    1.0    12.0            0.16            0.16            0.16   
12    1.0    13.0            0.01            0.01            0.00   
13    1.0    14.0            0.06            0.04            0.03   
14    1.0    15.0            0.08            0.06            0.04   
15    1.0    16.0            0.02            0.02            0.01   
16    1.0    18.0            0.08            0.08            0.08   
17    1.0    20.0            0.00            0.00            0.00   
18    1.0    21.0            0.06            0.08            0.10   
19    2.0     1.0            0.18            0.19            0.21   
20    2.0     2.0            0.08            0.08            0.08   

    %units_sold_Q4  
0             0.16  
1             0.06  
2             0.01  
3             0.00  
4             0.02  
5             0.01  
6             0.01  
7             0.02  
8             0.22  
9             0.01  
10            0.00  
11            0.15  
12            0.01  
13            0.08  
14            0.08  
15            0.02  
16            0.07  
17            0.00  
18            0.07  
19            0.18  
20            0.08

## Categories

In [58]:
df=df19
df['week'] = pd.to_datetime(df['week'])
df['quarter'] = df['week'].dt.quarter
df1=df[df['quarter']==1]
df2=df[df['quarter']==2]
df3=df[df['quarter']==3]
df4=df[df['quarter']==4]

In [60]:
# we will use a groupby to show for each store each Category and the percentage of units sold
quarter1 = df1.groupby(['store','Category']).agg({'  u':['sum']})
quarter1 = quarter1.groupby(level=0).apply(lambda x:x / float(x.sum())).round(2)
quarter1.columns = quarter1.columns.droplevel(1)

quarter2 = df2.groupby(['store','Category']).agg({'  u':['sum']})
quarter2 = quarter2.groupby(level=0).apply(lambda x:x / float(x.sum())).round(2)
quarter2.columns = quarter2.columns.droplevel(1)

quarter3 = df3.groupby(['store','Category']).agg({'  u':['sum']})
quarter3 = quarter3.groupby(level=0).apply(lambda x: x / float(x.sum())).round(2)
quarter3.columns = quarter3.columns.droplevel(1)

quarter4 = df4.groupby(['store','Category']).agg({'  u':['sum']})
quarter4 = quarter4.groupby(level=0).apply(lambda x: x / float(x.sum())).round(2)
quarter4.columns = quarter4.columns.droplevel(1)

quarter1=quarter1.reset_index()
quarter2=quarter2.reset_index()
quarter3=quarter3.reset_index()
quarter4=quarter4.reset_index()

In [62]:
#as the store repeats 21 times as the number of families, in order to merge the different quarters 
#we need first to create a unique index.

quarter1['Store_cat']= quarter1['store'].astype(str)+ '.' + quarter1['Category'].astype(str)
quarter2['Store_cat']= quarter2['store'].astype(str)+ '.' + quarter2['Category'].astype(str)
quarter3['Store_cat']= quarter3['store'].astype(str)+ '.' + quarter3['Category'].astype(str)
quarter4['Store_cat']= quarter4['store'].astype(str)+ '.' + quarter4['Category'].astype(str)

quarter2.drop(columns = ['store', 'Category'], inplace = True)
quarter3.drop(columns = ['store', 'Category'], inplace = True)
quarter4.drop(columns = ['store', 'Category'], inplace = True)

In [65]:
marketsharecat19=quarter1.merge(quarter2,how='outer', on='Store_cat', suffixes=('_Q1','_Q2'))
marketsharecat19=marketsharecat19.merge(quarter3,how='outer', on='Store_cat')
marketsharecat19=marketsharecat19.merge(quarter4, how='outer', on='Store_cat')
marketsharecat19.drop(columns = ['Store_cat'], inplace = True)

#change columns name 
marketsharecat19.columns = ['store', 'Category', '%units_sold_Q1', '%units_sold_Q2', '%units_sold_Q3', '%units_sold_Q4']

Let's do an example with store 292: we will see the top 20 categories sold for quarter 1

In [68]:
example=marketsharecat19[marketsharecat19['store']==292.0]
example.sort_values(by=['%units_sold_Q1'], ascending=False).head(20)

store  Category  %units_sold_Q1  %units_sold_Q2  %units_sold_Q3  \
26907  292.0     178.0            0.58            0.66            0.70   
26810  292.0      27.0            0.08            0.04            0.03   
26905  292.0     175.0            0.04            0.04            0.04   
26806  292.0      22.0            0.03            0.04            0.04   
26792  292.0       3.0            0.03            0.01            0.01   
26791  292.0       2.0            0.03            0.02            0.01   
26811  292.0      28.0            0.03            0.01            0.01   
26798  292.0      11.0            0.02            0.01            0.01   
26903  292.0     172.0            0.01            0.01            0.01   
26894  292.0     157.0            0.01            0.01            0.01   
26873  292.0     127.0            0.01            0.01            0.01   
26857  292.0     109.0            0.01            0.01            0.01   
26823  292.0      44.0            0.01            0.00            0.01   
26820  292.0      39.0            0.01            0.01            0.01   
26815  292.0      34.0            0.01            0.01            0.01   
26812  292.0      31.0            0.01            0.00            0.00   
26790  292.0       1.0            0.01            0.00            0.00   
26797  292.0      10.0            0.01            0.01            0.00   
26804  292.0      18.0            0.01            0.01            0.00   
26800  292.0      13.0            0.01            0.00            0.00   

       %units_sold_Q4  
26907            0.64  
26810            0.07  
26905            0.04  
26806            0.04  
26792            0.02  
26791            0.02  
26811            0.02  
26798            0.01  
26903            0.01  
26894            0.00  
26873            0.01  
26857            0.01  
26823            0.01  
26820            0.01  
26815            0.01  
26812            0.01  
26790            0.01  
26797            0.01  
26804            0.01  
26800            0.00

Category 178 takes 58% of the total market, followed by cat 27.

### 2018

In [80]:
df=df18
df['week'] = pd.to_datetime(df['week'])
df['quarter'] = df['week'].dt.quarter
df1=df[df['quarter']==1]
df2=df[df['quarter']==2]
df3=df[df['quarter']==3]
df4=df[df['quarter']==4]

In [82]:
# we will use a groupby to show for each store each Category and the percentage of units sold
quarter1 = df1.groupby(['store','Category']).agg({'  u':['sum']})
quarter1 = quarter1.groupby(level=0).apply(lambda x:x / float(x.sum())).round(2)
quarter1.columns = quarter1.columns.droplevel(1)

quarter2 = df2.groupby(['store','Category']).agg({'  u':['sum']})
quarter2 = quarter2.groupby(level=0).apply(lambda x:x / float(x.sum())).round(2)
quarter2.columns = quarter2.columns.droplevel(1)

quarter3 = df3.groupby(['store','Category']).agg({'  u':['sum']})
quarter3 = quarter3.groupby(level=0).apply(lambda x: x / float(x.sum())).round(2)
quarter3.columns = quarter3.columns.droplevel(1)

quarter4 = df4.groupby(['store','Category']).agg({'  u':['sum']})
quarter4 = quarter4.groupby(level=0).apply(lambda x: x / float(x.sum())).round(2)
quarter4.columns = quarter4.columns.droplevel(1)

quarter1=quarter1.reset_index()
quarter2=quarter2.reset_index()
quarter3=quarter3.reset_index()
quarter4=quarter4.reset_index()

#as the store repeats 21 times as the number of families, in order to merge the different quarters 
#we need first to create a unique index.

quarter1['Store_cat']= quarter1['store'].astype(str)+ '.' + quarter1['Category'].astype(str)
quarter2['Store_cat']= quarter2['store'].astype(str)+ '.' + quarter2['Category'].astype(str)
quarter3['Store_cat']= quarter3['store'].astype(str)+ '.' + quarter3['Category'].astype(str)
quarter4['Store_cat']= quarter4['store'].astype(str)+ '.' + quarter4['Category'].astype(str)

quarter2.drop(columns = ['store', 'Category'], inplace = True)
quarter3.drop(columns = ['store', 'Category'], inplace = True)
quarter4.drop(columns = ['store', 'Category'], inplace = True)

In [83]:
marketsharecat18=quarter1.merge(quarter2,how='outer', on='Store_cat', suffixes=('_Q1','_Q2'))
marketsharecat18=marketsharecat18.merge(quarter3,how='outer', on='Store_cat')
marketsharecat18=marketsharecat18.merge(quarter4, how='outer', on='Store_cat')
marketsharecat18.drop(columns = ['Store_cat'], inplace = True)

#change columns name 
marketsharecat18.columns = ['store', 'Category', '%units_sold_Q1', '%units_sold_Q2', '%units_sold_Q3', '%units_sold_Q4']
marketsharecat18.head(20)

store  Category  %units_sold_Q1  %units_sold_Q2  %units_sold_Q3  \
0     1.0       1.0            0.01            0.01            0.00   
1     1.0       2.0            0.02            0.01            0.01   
2     1.0       3.0            0.05            0.02            0.01   
3     1.0       4.0            0.00            0.00            0.00   
4     1.0       5.0            0.00            0.00            0.00   
5     1.0       7.0            0.01            0.00            0.00   
6     1.0       9.0            0.00            0.00            0.00   
7     1.0      10.0            0.01            0.01            0.00   
8     1.0      11.0            0.02            0.01            0.01   
9     1.0      13.0            0.00            0.00            0.00   
10    1.0      14.0            0.00            0.00            0.00   
11    1.0      16.0            0.00            0.00            0.00   
12    1.0      18.0            0.01            0.00            0.00   
13    1.0      21.0            0.00            0.00            0.00   
14    1.0      22.0            0.02            0.02            0.02   
15    1.0      23.0            0.00            0.00            0.00   
16    1.0      26.0            0.00            0.00            0.00   
17    1.0      27.0            0.06            0.03            0.02   
18    1.0      28.0            0.02            0.01            0.01   
19    1.0      31.0            0.00            0.00            0.00   

    %units_sold_Q4  
0             0.01  
1             0.02  
2             0.03  
3             0.00  
4             0.00  
5             0.01  
6             0.00  
7             0.01  
8             0.01  
9             0.00  
10            0.00  
11             NaN  
12            0.01  
13            0.00  
14            0.02  
15            0.00  
16            0.00  
17            0.05  
18            0.02  
19            0.00

### 2017

In [84]:
df=df17
df['week'] = pd.to_datetime(df['week'])
df['quarter'] = df['week'].dt.quarter
df1=df[df['quarter']==1]
df2=df[df['quarter']==2]
df3=df[df['quarter']==3]
df4=df[df['quarter']==4]

In [85]:
# we will use a groupby to show for each store each Category and the percentage of units sold
quarter1 = df1.groupby(['store','Category']).agg({'  u':['sum']})
quarter1 = quarter1.groupby(level=0).apply(lambda x:x / float(x.sum())).round(2)
quarter1.columns = quarter1.columns.droplevel(1)

quarter2 = df2.groupby(['store','Category']).agg({'  u':['sum']})
quarter2 = quarter2.groupby(level=0).apply(lambda x:x / float(x.sum())).round(2)
quarter2.columns = quarter2.columns.droplevel(1)

quarter3 = df3.groupby(['store','Category']).agg({'  u':['sum']})
quarter3 = quarter3.groupby(level=0).apply(lambda x: x / float(x.sum())).round(2)
quarter3.columns = quarter3.columns.droplevel(1)

quarter4 = df4.groupby(['store','Category']).agg({'  u':['sum']})
quarter4 = quarter4.groupby(level=0).apply(lambda x: x / float(x.sum())).round(2)
quarter4.columns = quarter4.columns.droplevel(1)

quarter1=quarter1.reset_index()
quarter2=quarter2.reset_index()
quarter3=quarter3.reset_index()
quarter4=quarter4.reset_index()

#as the store repeats 21 times as the number of families, in order to merge the different quarters 
#we need first to create a unique index.

quarter1['Store_cat']= quarter1['store'].astype(str)+ '.' + quarter1['Category'].astype(str)
quarter2['Store_cat']= quarter2['store'].astype(str)+ '.' + quarter2['Category'].astype(str)
quarter3['Store_cat']= quarter3['store'].astype(str)+ '.' + quarter3['Category'].astype(str)
quarter4['Store_cat']= quarter4['store'].astype(str)+ '.' + quarter4['Category'].astype(str)

quarter2.drop(columns = ['store', 'Category'], inplace = True)
quarter3.drop(columns = ['store', 'Category'], inplace = True)
quarter4.drop(columns = ['store', 'Category'], inplace = True)

In [86]:
marketsharecat17=quarter1.merge(quarter2,how='outer', on='Store_cat', suffixes=('_Q1','_Q2'))
marketsharecat17=marketsharecat17.merge(quarter3,how='outer', on='Store_cat')
marketsharecat17=marketsharecat17.merge(quarter4, how='outer', on='Store_cat')
marketsharecat17.drop(columns = ['Store_cat'], inplace = True)

#change columns name 
marketsharecat17.columns = ['store', 'Category', '%units_sold_Q1', '%units_sold_Q2', '%units_sold_Q3', '%units_sold_Q4']
marketsharecat17.head(20)

store  Category  %units_sold_Q1  %units_sold_Q2  %units_sold_Q3  \
0     1.0       1.0            0.01            0.00            0.01   
1     1.0       2.0            0.02            0.01            0.01   
2     1.0       3.0            0.05            0.02            0.02   
3     1.0       4.0            0.00            0.00            0.00   
4     1.0       5.0            0.00            0.00            0.00   
5     1.0       6.0            0.00            0.00             NaN   
6     1.0       7.0            0.01            0.00            0.00   
7     1.0       9.0            0.00            0.00            0.00   
8     1.0      10.0            0.01            0.00            0.01   
9     1.0      11.0            0.02            0.01            0.01   
10    1.0      12.0            0.00             NaN             NaN   
11    1.0      13.0            0.00            0.00            0.00   
12    1.0      14.0            0.00            0.00            0.00   
13    1.0      15.0            0.00            0.00             NaN   
14    1.0      16.0            0.00            0.00             NaN   
15    1.0      18.0            0.01            0.00            0.00   
16    1.0      21.0            0.00            0.00            0.00   
17    1.0      22.0            0.02            0.02            0.02   
18    1.0      23.0            0.00            0.00            0.00   
19    1.0      26.0            0.00            0.00            0.00   

    %units_sold_Q4  
0             0.01  
1             0.02  
2             0.04  
3             0.00  
4             0.00  
5              NaN  
6             0.01  
7             0.00  
8             0.01  
9             0.02  
10             NaN  
11            0.00  
12            0.00  
13            0.00  
14            0.00  
15            0.01  
16            0.00  
17            0.02  
18            0.00  
19            0.00

### 2016

In [87]:
df=df16
df['week'] = pd.to_datetime(df['week'])
df['quarter'] = df['week'].dt.quarter
df1=df[df['quarter']==1]
df2=df[df['quarter']==2]
df3=df[df['quarter']==3]
df4=df[df['quarter']==4]

In [88]:
# we will use a groupby to show for each store each Category and the percentage of units sold
quarter1 = df1.groupby(['store','Category']).agg({'  u':['sum']})
quarter1 = quarter1.groupby(level=0).apply(lambda x:x / float(x.sum())).round(2)
quarter1.columns = quarter1.columns.droplevel(1)

quarter2 = df2.groupby(['store','Category']).agg({'  u':['sum']})
quarter2 = quarter2.groupby(level=0).apply(lambda x:x / float(x.sum())).round(2)
quarter2.columns = quarter2.columns.droplevel(1)

quarter3 = df3.groupby(['store','Category']).agg({'  u':['sum']})
quarter3 = quarter3.groupby(level=0).apply(lambda x: x / float(x.sum())).round(2)
quarter3.columns = quarter3.columns.droplevel(1)

quarter4 = df4.groupby(['store','Category']).agg({'  u':['sum']})
quarter4 = quarter4.groupby(level=0).apply(lambda x: x / float(x.sum())).round(2)
quarter4.columns = quarter4.columns.droplevel(1)

quarter1=quarter1.reset_index()
quarter2=quarter2.reset_index()
quarter3=quarter3.reset_index()
quarter4=quarter4.reset_index()

#as the store repeats 21 times as the number of families, in order to merge the different quarters 
#we need first to create a unique index.

quarter1['Store_cat']= quarter1['store'].astype(str)+ '.' + quarter1['Category'].astype(str)
quarter2['Store_cat']= quarter2['store'].astype(str)+ '.' + quarter2['Category'].astype(str)
quarter3['Store_cat']= quarter3['store'].astype(str)+ '.' + quarter3['Category'].astype(str)
quarter4['Store_cat']= quarter4['store'].astype(str)+ '.' + quarter4['Category'].astype(str)

quarter2.drop(columns = ['store', 'Category'], inplace = True)
quarter3.drop(columns = ['store', 'Category'], inplace = True)
quarter4.drop(columns = ['store', 'Category'], inplace = True)

In [89]:
marketsharecat16=quarter1.merge(quarter2,how='outer', on='Store_cat', suffixes=('_Q1','_Q2'))
marketsharecat16=marketsharecat16.merge(quarter3,how='outer', on='Store_cat')
marketsharecat16=marketsharecat16.merge(quarter4, how='outer', on='Store_cat')
marketsharecat16.drop(columns = ['Store_cat'], inplace = True)

#change columns name 
marketsharecat16.columns = ['store', 'Category', '%units_sold_Q1', '%units_sold_Q2', '%units_sold_Q3', '%units_sold_Q4']
marketsharecat16.head(20)

store  Category  %units_sold_Q1  %units_sold_Q2  %units_sold_Q3  \
0     1.0       1.0            0.01            0.01            0.01   
1     1.0       2.0            0.02            0.01            0.01   
2     1.0       3.0            0.04            0.03            0.02   
3     1.0       4.0            0.00            0.00            0.00   
4     1.0       5.0            0.00            0.00            0.00   
5     1.0       6.0            0.00            0.00            0.00   
6     1.0       7.0            0.01            0.01            0.00   
7     1.0       9.0            0.00            0.00            0.00   
8     1.0      10.0            0.01            0.01            0.01   
9     1.0      11.0            0.02            0.02            0.01   
10    1.0      12.0            0.00            0.00            0.00   
11    1.0      13.0            0.00            0.00            0.00   
12    1.0      14.0            0.00            0.00            0.00   
13    1.0      15.0            0.00            0.00            0.00   
14    1.0      16.0            0.00            0.00            0.00   
15    1.0      18.0            0.01            0.01            0.00   
16    1.0      21.0            0.00            0.00            0.00   
17    1.0      22.0            0.02            0.02            0.02   
18    1.0      23.0            0.00            0.00            0.00   
19    1.0      26.0            0.00            0.00            0.00   

    %units_sold_Q4  
0             0.01  
1             0.02  
2             0.04  
3             0.00  
4             0.00  
5             0.00  
6             0.01  
7             0.00  
8             0.01  
9             0.02  
10            0.00  
11            0.00  
12            0.00  
13            0.00  
14            0.00  
15            0.01  
16            0.00  
17            0.02  
18            0.00  
19            0.00